### PR TITLE
Feature gate defaulted type parameters appearing outside of types

### PR DESCRIPTION
**Pull request by nikomatsakis**
It was recently realized that we accept defaulted type parameters everywhere, without feature gate, even though the only place that we really _intended_ to accept them were on types. This PR adds a lint warning unless the "type-parameter-defaults" feature is enabled. This should eventually become a hard error.

This is a [breaking-change] in that new feature gates are required (or simply removing the defaults, which is probably a better choice as they have little effect at this time). Results of a [crater run](https://gist.github.com/nikomatsakis/760c6a67698bd24253bf) suggest that approximately 5-15 crates are affected. I didn't do the measurement quite right so that run cannot distinguish "true" regressions from "non-root" regressions, but even the upper bound of 15 affected crates seems relatively minimal. 

cc @rust-lang/lang 
r? @pnkfelix

### CODE REVIEW

File: src/librustc/lint/context.rs [hunk 1]
```diff
@@ -413,6 +415,18 @@ pub fn raw_struct_lint<'a>(sess: &'a Session,
         _ => sess.bug("impossible level in raw_emit_lint"),
     };
 
+    // Check for future incompatibility lints and issue a stronger warning.
+    let future_incompat_lints = &lints.lint_groups[builtin::FUTURE_INCOMPATIBLE];
+    let this_id = LintId::of(lint);
+    if future_incompat_lints.0.iter().any(|&id| id == this_id) {
+        let msg = "this lint will become a HARD ERROR in a future release!";
```
Comment (pnkfelix) [inline src/librustc/lint/context.rs]: ah cool, good to CSE this messaging in one place.

File: src/test/compile-fail/type-parameter-invalid-lint.rs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![deny(future_incompatible)]
+#![allow(dead_code)]
+
+fn avg<T=i32>(_: T) {}
+//~^ ERROR defaults for type parameters are only allowed
+//~| NOTE HARD ERROR
```
Comment (pnkfelix) [inline src/test/compile-fail/type-parameter-invalid-lint.rs]: (not so thrilled about this impoverished "NOTE", just in terms of what future readers will think, but then again this test will have to change "soon", as in after the warning cycle is done, so who cares!)

**Comment by pnkfelix** [comment]
(graar i hate githubs presentation of the commit series)

**Comment by pnkfelix** [comment]
(I suspect we should not actually take any special effort to backport this to beta. But I'll leave the nominated tag to ensure that the team discusses it.)

**Comment by bluss** [comment]
I object to removing defaults in `trait` definitions. The description does not mention it either way, so I'm not 100% sure which way you are going here.

Something like the following is working fine for me today. It also allows backwards compatible addition of type parameters just like for structs.

```rust
trait IndexRange<T=usize> { fn start(&self) -> Option<T> }

fn index<R: IndexRange>(r: R) { ... }
```

**Comment by petrochenkov** [comment]
@bluss
Defaults are still permitted on traits (as well as on impls, type aliases, enums and structs).
This PR restricts their use on functions and methods (and, technically, foreign functions).

**Comment by bluss** [comment]
@petrochenkov Ok, that's great. Some of the messages need updating for that.

**Comment by m4rw3r** [comment]
Some functional programming with generics would benefit from having the default type parameters, provided that `type-parameter-defaults` is enabled. It would enable for eg. skipping error types or letting the user skip type annotations when functions would automatically convert types if no conversion is necessary (see Chomp's [`ret`](https://github.com/m4rw3r/chomp/blob/2c4469c22c84ed60d292d8b479c1d5be35b640ef/src/input.rs#L124) or [`bind`](https://github.com/m4rw3r/chomp/blob/2c4469c22c84ed60d292d8b479c1d5be35b640ef/src/parse_result.rs#L142) for examples).

Since it is not yet enabled by default I will follow the recommendation and remove these defaults from Chomp.

**Comment by bluss** [comment]
type parameter defaults seem to be going rather than coming to rust, what it looks like now, @m4rw3r

**Comment by bluss** [comment]
This should link to the tracking issue for default type parameter fallback: #27336  (now it does)

**Pull request merged**